### PR TITLE
Add op_idx to par info

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**

Currently it is not possible to get the global index of a trainable operation in a tape.

**Description of the Change:**

This PR updates `_par_info` and adds a method `_get_operation` which adds the operation index to return. This private function will replace get_operation after a deprecation cycle.

**Benefits:**

Possible to obtain the position of a trainable operation in a quantum circuit.

### CODE REVIEW

File: tests/tape/test_qscript.py
Comment (rmoyard) [inline tests/tape/test_qscript.py:296]:
```suggestion
```

**Comment by timmysilv** [comment]
I wonder if there's a nice public name we can use like `get_op`, so that when we deprecate `get_operation`, we won't have to magically re-introduce it. It would be weird to have a release saying `get_operation` is deprecated, and then in the next release, poof! there it is again, but with a new return value to unpack. another idea would be to add a kwarg like `global_id=False`, and deprecate the `False` (default behaviour), flip it to `True` by default in the next release, then remove the kwarg altogether after that. that process would be slightly painful.. but it's an option

File: pennylane/gradients/hamiltonian_grad.py
Comment (dwierichs) [inline pennylane/gradients/hamiltonian_grad.py:27]: Wouldn't it be more consistent to use the new keyword argument here, and in particular its default? :thinking:

Comment (dwierichs) [reply pennylane/gradients/hamiltonian_grad.py:27]: Same for the changes in `parameter_shift.py`

File: pennylane/tape/qscript.py
Comment (dwierichs) [inline pennylane/tape/qscript.py:597]: As dictionaries are guaranteed to have a fixed order (the input order), we could do 
```suggestion
        return list(info.values())
```

File: tests/tape/test_qscript.py
Comment (dwierichs) [inline tests/tape/test_qscript.py:201]:
```suggestion
        """Tests the function get_operation."""
```

Comment (dwierichs) [reply tests/tape/test_qscript.py:201]: Wait, this test doesn't seem to call `get_operation`, does it?

File: tests/tape/test_qscript.py
Comment (dwierichs) [inline tests/tape/test_qscript.py:239]: This only tests that the warnings is raised once. Is that sufficient?

**Review by dwierichs — COMMENTED** [review]
Not replacing any of the requested reviews, I'm just keen on this becoming a feature! :)

File: pennylane/gradients/hamiltonian_grad.py
Comment (albi3ro) [reply pennylane/gradients/hamiltonian_grad.py:27]: +1.  We don't want to be relying on private methods outside of the class itself.

File: pennylane/tape/qscript.py
Comment (albi3ro) [inline pennylane/tape/qscript.py:560]: ? `idx` is "the trainable parameter idx", so how is `return_op_index` also "the trianable parameter idx?

File: pennylane/tape/qscript.py
Comment (albi3ro) [reply pennylane/tape/qscript.py:597]: If we ever put the time into cleaning all this behaviour up, we could even have `_par_info` be a list of named tuples.

File: pennylane/tape/qscript.py
Comment (albi3ro) [inline pennylane/tape/qscript.py:580]: Does this really need to get extracted into a separate method? Especially as all the code can be written on one line: 
```
tuple(self._par_info[self.trainable_params[idx]].values())
```

File: tests/tape/test_qscript.py
Comment (albi3ro) [inline tests/tape/test_qscript.py:224]: This test has the exact same name, and thus won't be run.

File: tests/tape/test_qscript.py
Comment (albi3ro) [inline tests/tape/test_qscript.py:300]:
```suggestion
    def test_get_operation(self):
```

But the same name once again. For the name, include details about how this test *differs* from the other ones. What exact property of `get_operation` is it testing?

File: tests/tape/test_qscript.py
Comment (rmoyard) [reply tests/tape/test_qscript.py:201]: This test is a mistake of copy pasting, removed

File: tests/tape/test_qscript.py
Comment (rmoyard) [reply tests/tape/test_qscript.py:224]: Copy paste mistake.

File: tests/tape/test_qscript.py
Comment (rmoyard) [reply tests/tape/test_qscript.py:300]: Here it is testing the private method and this test will replace the `test_get_operation` in the future.

File: pennylane/tape/qscript.py
Comment (rmoyard) [reply pennylane/tape/qscript.py:560]: The function also returns the operation index.

File: tests/tape/test_qscript.py
Comment (rmoyard) [reply tests/tape/test_qscript.py:239]: I can add one more test for the warning.

File: tests/tape/test_qscript.py
Comment (albi3ro) [reply tests/tape/test_qscript.py]: How about `test_private_get_operation`?

Comment (rmoyard) [reply tests/tape/test_qscript.py]: Yes I already change it :+1:

File: pennylane/tape/qscript.py
Comment (albi3ro) [inline pennylane/tape/qscript.py:564]: We should update this as well.  Now it could also return `tuple[.Operator, int, int]` as well, depending on the `return_op_index` keyword argument.

File: doc/releases/changelog-dev.md
Comment (albi3ro) [inline doc/releases/changelog-dev.md:310]: When? Always a good idea to have a timeline for these things so we know when we need to go back and complete the deprecation cycle.

**Review by albi3ro — COMMENTED** [review]
Overall, this is a fairly internal change with a fairly specific set of use cases, so I'm not too concerned about any breaking impact.

Since this has a deprecation cycle over which the behaviour will be changed, we should be explicit about the timeline and add an entry to `doc/developement/deprecations.rst`.

I'm not entirely convinced we need a separate private method. Especially since it's basically just:
```
info["op"], info["op_idx"], info["p_idx"] if return_op_index else info["op"], info["p_idx"]
```
But this preference is not that important.

I'll approve once we have a time line and a deprecations page entry.

File: doc/releases/changelog-dev.md
Comment (rmoyard) [reply doc/releases/changelog-dev.md:310]: 0.29 introduced this change, 0.30 we switch

File: doc/releases/changelog-dev.md
Comment (timmysilv) [inline doc/releases/changelog-dev.md:308]: is this a breaking change because a warning is raised?

**Review by timmysilv — APPROVED** [review]
pending response to Christina's request

File: doc/releases/changelog-dev.md
Comment (rmoyard) [reply doc/releases/changelog-dev.md:308]: Partly that and just that users are aware, that method `get_operator` will have a different signature after `0.30.0`

Comment (timmysilv) [reply doc/releases/changelog-dev.md:308]: kk. not sure i'd strictly call this "breaking", but putting it here is probably the best way to let users know asap so lgtm

File: doc/development/deprecations.rst
Comment (dwierichs) [inline doc/development/deprecations.rst:12]:
```suggestion
  - The old signature is replaced with the new one in v0.30
```

File: doc/releases/changelog-dev.md
Comment (dwierichs) [inline doc/releases/changelog-dev.md:310]:
```suggestion
  the default in version 0.30.
```

File: doc/releases/changelog-dev.md
Comment (dwierichs) [reply doc/releases/changelog-dev.md:308]: I agree, it's not really a breaking change. Maybe we could have a "Deprecations" section in the changelog? :thinking: @albi3ro is there a standard way to announce deprecations in the changelog?

File: pennylane/gradients/hamiltonian_grad.py
Comment (dwierichs) [inline pennylane/gradients/hamiltonian_grad.py:27]: I see two versions for this change:
1. We do it like this now, and have to touch this line again in the version that removes the keyword argument.
2. We do not use the keyword argument here, i.e we do not have to touch this file at all. Then we just have to touch this line later when the default is switched to include the `_, ` part.

It's super minimal but I think option 2 is cleaner. Not a blocker I think.

Comment (dwierichs) [reply pennylane/gradients/hamiltonian_grad.py:27]: Same applies to the changes in `parameter_shift.py`

File: pennylane/tape/qscript.py
Comment (dwierichs) [inline pennylane/tape/qscript.py:560]:
```suggestion
            return_op_index (bool): Whether the function also returns the operation index.
```

File: pennylane/tape/qscript.py
Comment (dwierichs) [inline pennylane/tape/qscript.py:563]:
```suggestion
            tuple[.Operation, int, int]: tuple containing the corresponding
            operation, the operation index, and an integer representing the argument index,
```

**Review by dwierichs — APPROVED** [review]
Super tiny comments left from my side. Definitely wait for another approve to make sure, I'd say.

**Review by albi3ro — COMMENTED** [review]
I'm not going to block this PR, but I'm actually unconvinced this even needs to be a breaking change at all.

As I understand it, in a particular situation, we also need the index for the operator in the circuit. So we just need a way to be able to also request the index of the operator. I don't think there's a particular need to change the default behaviour.

**Pull request merged**
